### PR TITLE
First commit!

### PR DESCRIPTION
**Pull request by lewisoaten**
Created framework app connecting through from a React frontend, Rocket API, and SQLx DB.

This version simply selects the first record in a mock table and shows the ID in a plaintext response.

image

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      - id: no-commit-to-branch
+  - repo: https://github.com/pre-commit/mirrors-prettier
+    rev: v2.7.1
+    hooks:
+      - id: prettier
+  - repo: https://github.com/pre-commit/mirrors-eslint
+    rev: v8.23.1
+    hooks:
+      - id: eslint
+  - repo: https://github.com/doublify/pre-commit-rust
```
Comment (0atman) [inline .pre-commit-config.yaml:35]: make sure to dial clippy up to 11 (`pedantic nursery unwrap_used`)

File: api/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+FROM rust:1.63.0
+
+WORKDIR /usr/src/app/calandar
+
+EXPOSE 8000
+
+RUN cargo install cargo-watch
+RUN cargo install sqlx-cli
+
+COPY . .
```
Comment (0atman) [inline api/Dockerfile:10]: nb that cargo's cache is in your home directory. If recompiles are taking a toll on your life, use cargo locally.

File: api/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#[get("/")]
+async fn index(pool: &rocket::State<PgPool>) -> String {
+    // Make a simple query to return the given parameter (use a question mark `?` instead of `$1` for MySQL)
+    let test_id: i32 = match sqlx::query!("SELECT id FROM test")
+        .fetch_one(pool.inner())
+        .await
+    {
+        Ok(test_id) => test_id.id,
+        Err(e) => {
+            // One day we should handle errors better (HTTP return codes)
```
Comment (0atman) [inline api/src/main.rs:15]: Investigate bubbling up errors using `?`, see what rocket likes

File: api/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
+#[macro_use]
+extern crate rocket;
+use sqlx::postgres::PgPool;
+use sqlx::postgres::PgPoolOptions;
+
+#[get("/")]
+async fn index(pool: &rocket::State<PgPool>) -> String {
+    // Make a simple query to return the given parameter (use a question mark `?` instead of `$1` for MySQL)
+    let test_id: i32 = match sqlx::query!("SELECT id FROM test")
```
Comment (0atman) [inline api/src/main.rs:9]: Once you have structured data, you'll want to use sqlx's auto-destructuring into compile-checked types, check the documentation, it looks like this:

```rust
#[derive(sqlx::FromRow)]
struct User { name: String, id: i64 }

let mut stream = sqlx::query_as::<_, User>("SELECT * FROM users WHERE email = ? OR name = ?")
    .bind(user_email)
    .bind(user_name)
    .fetch(&mut conn);
```

**Review by 0atman — APPROVED** [review]
Very nice!

File: .pre-commit-config.yaml (hunk 1 above)
Comment (lewisoaten) [reply .pre-commit-config.yaml:35]: Thanks, will set that now.

File: api/Dockerfile (hunk 2 above)
Comment (lewisoaten) [reply api/Dockerfile:10]: OK, I'll investigate how to set the cache directory.

File: api/src/main.rs (hunk 3 above)
Comment (lewisoaten) [reply api/src/main.rs:15]: Yeah, played with this a lot. It's non-obvious how to elegantly marshall Rust Errors into HTTP Error codes. I'll leave this for now, but plan on doing more here soon.

File: api/Dockerfile (hunk 2 above)
Comment (lewisoaten) [reply api/Dockerfile:10]: I was able to change the `CARGO_HOME` directory, but it didn't work to improve recompilation time under most conditions. 

These docker image builds need an overhaul soon anyway, but that shouldn't block functions right now - so long as the inner loop is fast, which it is.

I'll come back to this down the line. Thanks for the suggestion

**Pull request merged**